File: Two_Pointers/6.Merge Two Sorted Arrays (Merge Sorted Array II)/Solution.java
public class Solution {
    /*
     * @param A: sorted integer array A
     * @param B: sorted integer array B
     * @return: A new sorted integer array
     */
    public int[] mergeSortedArray(int[] A, int[] B) {
        if (A == null || A.length == 0) {
            return B;
        }
        if (B == null || B.length == 0) {
            return A;
        }
 
        int[] res = new int[A.length + B.length];
        int idx = 0;
        int pa = 0;
 
        for(int i = 0; i < B.length; i++){
            int position = binarySearch(A, B[i]);
            while(pa < position){
                res[idx++] = A[pa++];
            }
            res[idx++] = B[i];
        }
 
        while(pa < A.length){
            res[idx++] = A[pa++];
        }
 
        return res;
    }
    
    private int binarySearch(int[] A, int target){
        int left = 0;
        int right = A.length-1;

        while(left <= right){
            int mid = left + (right-left)/2;
            if (A[mid] == target) {
                return mid;
            } else if (target < A[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
}

File: Two_Pointers/6.Merge Two Sorted Arrays (Merge Sorted Array II)/Solution.py
class Solution:
    """
    @param A: sorted integer array A
    @param B: sorted integer array B
    @return: A new sorted integer array
    """
    def mergeSortedArray(self, A, B):
        # write your code here
        if not A or len(A) == 0:
            return B
        if not B or len(B) == 0:
            return A

        if len(A) < len(B):
            temp = A
            A = B
            B = temp

        # binary search and insert
        for ele in B:
            left = 0
            right = len(A) - 1
            while left + 1 < right:
                mid = int(left + (right - left) / 2)
                if ele < A[mid]:
                    right = mid
                else:
                    left = mid

            if ele < A[left]:
                A.insert(left, ele)
            elif ele > A[right]:
                A.insert(right + 1, ele)
            else:
                A.insert(right, ele)

        return A
